docs/features/component-versioning.md

```diff
@@ -0,0 +1,88 @@
+
+Although blocked from loading, the current user experience of using a component targeting a newer compiler version is sub-par and opaque as to the action the
+user needs to take to resolve it.
+
+Instead, it would be ideal if the author had a way to specify a particular feature of the compiler that is required; older compilers (at least the
+ones going forward from now on) would be able to detect this feature requirement at load time, and specify a specific warning that this version of the compiler cannot load the component.
+
+## Proposal
+
+We introduce a new Attribute `[RoslynFeature(...)]` that a component author can apply to their component definition. The attribute would take a required string parameter, which is a comma separated list of features the component requires in order to load.
+
+`[RoslynFeature("Generator.PostInitialization, LanguageVersion9")]`
+
+_(__Alternative__: this could directly be an array, rather than separated list?)_
+
+These features would be defined in the compiler, and passed to the component loader. The component loader would check that the requirements are met, and issue a warning when unable to meet them. The requirements would be accessible via API to enable the author to strongly reference the requirement (either an Enum or set of readonly properties).
+
+The exact requirement list is TBD, but would likely include language version and specific component features we introduce going forward. These features could be dynamically calculated by the compiler: for instance, whilst still in preview a feature might only be enabled if the selected language version is also preview.
+
+While we'd allow authors to key on language version, we wouldn't be changing any of the existing functionality in the compiler to use these requirements: it would purely be for making decisions about whether to load the component or not.
+
+Features would also be able to represent super-sets of other features, for instance `LanguageVersion9` would imply `LanguageVersion1..8`. Components without the attribute would always be unconditionally loaded, with the same behavior as today.
+
+### Selective disabling
+
+A second use case for this feature would be to selective _disable_ the loading of certain components.
+
+Consider the migration of an `ISourceGenerator` to `IIncrementalGenerator`: the author may wish to create a new `IIncrementalGenerator`, but still fall back to the original `ISourceGenerator` in older compilers. The above proposal allows the user to *enable* the incremental generator when supported, but conversely they also need a mechanism to *disable* the `ISourceGenerator`.
+
+The `[RoslynFeature]` attribute would support a second, optional parameter that would cause the component to **not** load when the given feature is present.
+
+`[RoslynFeature("", disableWhen: "IncrementalGenerator")]`
```

Comment:
```suggestion
`[RoslynFeature("", disableWhen: "IncrementalGenerator")]`
```

Make the required set optional? The "" is a bit unclear what it might mean if you haven't read this doc.